src/util/heap/blockpageresource.rs

```diff
@@ -0,0 +1,417 @@
+use super::pageresource::{PRAllocFail, PRAllocResult};
+use super::PageResource;
+use crate::util::address::Address;
+use crate::util::constants::*;
+use crate::util::conversions::bytes_to_pages;
+use crate::util::heap::layout::heap_layout::VMMap;
+use crate::util::heap::layout::vm_layout_constants::*;
+use crate::util::heap::pageresource::CommonPageResource;
+use crate::util::heap::space_descriptor::SpaceDescriptor;
+use crate::util::opaque_pointer::*;
+use crate::vm::*;
+use atomic::{Atomic, Ordering};
+use spin::RwLock;
+use std::cell::UnsafeCell;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Mutex;
+
+const UNINITIALIZED_WATER_MARK: i32 = -1;
+const LOCAL_BUFFER_SIZE: usize = 128;
+
+/// A fast PageResource for fixed-size block allocation only.
+pub struct BlockPageResource<VM: VMBinding> {
```

Comment:
Make it clear that the current implementation only support contiguous memory. And either make it clear why discontiguous cannot be supported, or add a `TODO` for discontiguous memory.

Reply:
I am also wondering if we can internally use `FreeListPageResource` or `MonotonePageResource` for allocating new pages. So when we allocate, we try get from the block pool first. If that fails, we get from other page resource. In that case, we do not need to worry about contiguous/discontiguous.

Reply:
Switched to `FreeListPageResource`